torch/csrc/jit/tensorexpr/external_functions.cpp

```diff
@@ -181,6 +183,66 @@ void nnc_aten_triangular_solve(
   }
 }
 
+#ifdef USE_XNNPACK
+
+void nnc_prepacked_linear_clamp_run(
+    int64_t bufs_num,
+    void** buf_data,
+    int64_t* buf_ranks,
+    int64_t* buf_dims,
+    int8_t* buf_dtypes,
+    int64_t args_num,
+    int64_t* extra_args) {
+  using namespace at::native::xnnpack;
+
+  std::vector<at::Tensor> tensors =
+      constructTensors(bufs_num - 1, buf_data, buf_ranks, buf_dims, buf_dtypes);
+
+  const at::Tensor& x = tensors[1];
+  const auto prepacked =
+      c10::intrusive_ptr<LinearOpContext>::unsafe_reclaim_from_nonowning(
+          reinterpret_cast<LinearOpContext*>(buf_data[2]));
+
+  static auto op =
+      c10::Dispatcher::singleton()
```

Comment:
Why do you need to go through dispatcher. I suppose in a generic case you do need to know which op to invoke?

Reply:
> Why do you need to go through dispatcher. I suppose in a generic case you do need to know which op to invoke?

@kimishpatel my original thought was to keep using dispatcher before we figure out how to call third-party/xnnpack library directly without going through the aten/native/xnnpack wrappers. I didn't want to have much dependency on aten/native/... stuff directly as they are generally considered "hidden" implementation details.

For clients like static runtime where we don't want to incur dispatcher overhead, I think we generate separate public header functions for overhead-avoidance purpose. Unfortunately we don't have equivalent thing for manually registered custom ops - I don't want to set a "bad" example here :)

After inspecting the code a bit more, I changed the plan a little bit. Seems these XNNPack ops are very special - they simply call the "context->run(input)" without doing any other pre/post processing. The existing pattern is to wrap all the pre/post processing logic inside the LinearOpContext/Conv2dOpContext classes, which we have to expose anyway. So here I changed it to call "context->run()" directly as well. The risk is that if we ever change the aten side linear_clamp_run() / conv2d_clamp_run() implementation to do more work than "context->run()", then we need to update these external function wrappers here as well. In other words, "LinearOpContext" and "Conv2dOpContext" methods become "public" APIs that the op developers need to be aware of.